arbor/include/arbor/communication/remote.hpp

```diff
@@ -0,0 +1,182 @@
+    float time = 0.0f;
+};
+
+using ctrl_message = std::variant<msg_null,
+                                  msg_abort,
+                                  msg_epoch,
+                                  msg_done>;
+
+// Exceptions
+struct remote_error: std::runtime_error {
+    remote_error(const std::string& msg): std::runtime_error{msg} {}
+};
+
+struct unexpected_version: remote_error {
+    unexpected_version(): remote_error{"Arbor remote: Magic or Version mismatch."} {}
+};
+
+struct unexpected_message: remote_error {
+    unexpected_message(): remote_error{"Arbor remote: Received unknown tag."} {}
+};
+
+struct mpi_error: remote_error {
+    mpi_error(const std::string& where, int rc):
+        remote_error{"MPI failed in " + where + " with error=" + std::to_string(rc)} {}
+};
+
+inline
+void mpi_checked(int rc, const std::string& where) {
+    if (rc != MPI_SUCCESS) throw mpi_error{where, rc};
+}
+
+// Prepare byte buffer for sending.
+inline
+std::vector<char> make_send_buffer(const ctrl_message& msg, std::vector<char>& send) {
+    std::size_t ptr = 0;
+    send[ptr++] = ARB_REMOTE_MAGIC;
+    send[ptr++] = ARB_REMOTE_VERSION_MAJOR;
+    send[ptr++] = ARB_REMOTE_VERSION_MINOR;
+    send[ptr++] = ARB_REMOTE_VERSION_PATCH;
+    auto visitor = [&send, &ptr] (auto&& m) {
+        using T = std::decay_t<decltype(m)>;
+        send[ptr++] = T::tag;
+        memcpy(send.data() + ptr, &m, sizeof(m));
+    };
+    std::visit(visitor, msg);
+    return send;
+}
+
+// Consume receive buffer and generate a message
+inline
+ctrl_message make_ctrl_message(const std::vector<char>& recv) {
+    std::size_t ptr = 0;
+    std::uint8_t mag = recv[ptr++];
+    std::uint8_t maj = recv[ptr++];
+    std::uint8_t min = recv[ptr++];
+    std::uint8_t pat = recv[ptr++];
+    if ((mag != ARB_REMOTE_MAGIC) ||
+        (maj != ARB_REMOTE_VERSION_MAJOR) ||
+        (min != ARB_REMOTE_VERSION_MINOR) ||
+        (pat != ARB_REMOTE_VERSION_PATCH)) throw unexpected_message{};
+    std::uint8_t tag = recv[ptr++];
+    auto payload = recv.data() + ptr;
+    ctrl_message result = msg_null{};
+    switch(tag) {
+        case msg_null::tag: {
+            result = *reinterpret_cast<const msg_null*>(payload);
+            break;
+        }
+        case msg_abort::tag: {
+            result = *reinterpret_cast<const msg_abort*>(payload);
+            break;
+        }
+        case msg_epoch::tag: {
+            result = *reinterpret_cast<const msg_epoch*>(payload);
+            break;
+        }
+        case msg_done::tag: {
+            result = *reinterpret_cast<const msg_done*>(payload);
+            break;
+        }
+        default: throw unexpected_message{};
+    }
+    return result;
+}
+
+// Exchange control message. NOTE: This makes clever (?) use of MPI_Allreduce
+// to achieve simulatenous duplex mode broadcast by preparing zeroed buffers
+// on all ranks except the root and then adding up tons of zeros with a single
+// payload.
+inline
```

Comment:
please explain why this is better than a regular broadcast

Reply:
It's two broadcasts in duplex, ie assume group 0 has data `[42, 0, 0]` and group `[23, 0]` (42 and 23 are stand-ins for our payloads while the 0s shall represent our zeroed buffers). After the `Allreduce(..., inter)` we'll get `[23, 23, 23]` and `[42, 42]` in groups 0 and 1. If we were using broadcasts, we'd need to take care with ordering, ie who goes first group=0:rank=0 or group=1:rank=0. Thus we'd have to offer two different functions, explain who uses what when, ... This solution is a bit trippy, but it's one call, no ordering :D

Reply:
ah I wasn't aware that the intercomm broadcast is *not* duplex, and therefore has asymmetric function arguments. I see, very clever. But more expensive. Could  we do a sendrecv root to root and then local broadcast instead?

Reply:
Yes, but if that's more efficient?! Allreduce usually is one of the core functions to optimise, it's part of the acceptance test for most major installations :D